### DISCUSSION TITLE
Can you use bt.optimize for things like trade size, stop loss size, margin, etc.?

### DISCUSSION THREAD

**Discussion by CadeHalcyon** [kernc/backtesting.py#799] [General]
Sorry if this has already been answered, I have trawled through the previous discussions and could not find a definitive answer... essentially, is it possible to define things like stop loss size, trade size (e.g. as % of balance), leverage/margin etc. Class variables and use them with bt.optimize?  Thanks

**Comment by IndiasFernandes** [comment]
Yes, it is possible to use everything. You can also optimize any of those parameters. All the info is in the documentation.

For example, in the strategy, if you open a new trade you can use the following:

self.buy()

in order to set a stop-loss and a trade size you should use:

self.buy(tp={your take profit price value}, size={your trade size})

More info here:
https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Order

**Reply by CadeHalcyon** [reply]
Thanks, I have added the parameters and they run OK via run, but when I enter a range of values to test in optimize I get the following error msg:
`class VA_Strategy(Strategy):

stop_loss = 0.22
stats = bt.optimize(stop_loss = range(0.01, 0.25, 0.01), maximize='Return')`

`---> 69 stats = bt.optimize(stop_loss = range(0.01,0.25,0.01), maximize='Return')
TypeError: 'float' object cannot be interpreted as an integer`

**Reply by kernc** [reply]
Seems like [range doesn't handle floats](https://www.google.com/search?q=TypeError%3A+'float'+object+cannot+be+interpreted+as+an+integer). Try `np.linspace(...).tolist()`.

**Reply by CadeHalcyon** [reply]
I had a go with this but sadly can't get to work:

`AttributeError                            Traceback (most recent call last)
Input In [143], in <cell line: 69>()
     65 bt = Backtest(df_final, VA_Strategy, cash=1000, 
     66               commission=0.0012, exclusive_orders=True, trade_on_close = True, margin=0.333)#margin=.125,
     67 #stats = bt.run() #run this if not using optimize
---> 69 stats = bt.optimize(stop_loss = np.linspace(start=0.01, stop=0.25, num=26, retstep=0.01).tolist(), maximize='Return [%]')
     71 bt.plot(resample=True) #set resample to False if want original candle size and not worried about 10,000 limit
     72 print(stats)

AttributeError: 'tuple' object has no attribute 'tolist'
`
Then tried dropping the "tolist.()":

`---> 69 stats = bt.optimize(stop_loss = np.linspace(start=0.01, stop=0.25, num=26, retstep=0.01), maximize='Return [%]')

TypeError: unhashable type: 'numpy.ndarray'
`

What's odd is that I have watched videos by Algovibes and Chad Thackeray who both use range and use numbers for SMA periods and/or upper and lower limits for RSIs and don't seem to have this issue.

``

**Reply by CadeHalcyon** [reply]
Ah, me being silly, theirs worked because using integers, if using numbers less than 1 for stop losses I have found `range(0.01, 0.25, 0.01)` can be passed using numpy `np.arange(0.01, 0.25, 0.01)`.  This worked:

`stats = bt.optimize(stop_loss = np.arange(0.1, 0.25, 0.01).tolist(), maximize='Return [%]')`

Once you know, you know!

**Reply by CadeHalcyon** [reply]
One more thing... is it possible to use optimize for the margin?  I couldn't seem to work out how to do this.  Thank you.
